### PR TITLE
test: add IBC test

### LINKED ISSUE #5387: Tastora IBC Test
### Overview

Ensure IBC functionality works correctly with Tastora

### Test Case

- Deploy App vN and a Second Chain
- Establish Connection/Channel with the Hermes relayer
- Send tokens back and forth (can also verify token filter works here)
- Upgrade to N + 1
- Send tokens back and forth on same channel
- Create new channel and Send tokens back and forth.

Dependant on https://github.com/celestiaorg/tastora/issues/77

### PR DESCRIPTION
**Pull request by chatton**
## Overview

This PR adds an IBC test verifying outbound ibc transfers (it can be extended in a followup to send them back, I left it out for now as the PR was already growing quite a bit)

We bump to the newest version of tastora which has support for ibc and hermes.

The test introduced performs the following steps

- deploys 2 chains
- deploys hermes
- create an IBC connection and channel
- performs a token transfer
- upgrades from app version N-1 to N
- performs token transfer over the same channel
- creates a new connection and channel after upgrade
- performs token transfer over a new channel

A follow up PR can make sure the tokens can be sent back over the channels and are unwound into the native denom.

There is one hack/cavet with how the second chain is being deployed.

Because the global SDK config is sealed and the bech32 prefix is "celestia", some SDK functions which depend on the global type explode when dealing with a chain that has a different prefix e.g. "cosmos", what I have done for now is created a modified version of the dummy simapp from the ibc-go repo and changed the prefix to "celestia" to work around this. (it lives at `ghcr.io/chatton/ibc-go-simd:v8.5.0` build and pushed locally from [this branch](https://github.com/chatton/ibc-go/tree/release/v8.5.x)) 

It currently is sitting in my ghcr registry, this can either be moved into the celestiaorg ghcr registry, or, ideally, we find a workaround to enable different bech32 prefixes on chains.

This would likely involve "unsealing" or conditionally sealing the global config in celestia-app which doesn't seam ideal either.

This test still successfully verifies ibc transfers since it technically doesn't matter if the bech32 prefix is the same or not.

closes #5387

### CODE REVIEW

File: test/docker-e2e/dockerchain/testchain.go [hunk 1]
```diff
@@ -51,15 +52,22 @@ func NewCelestiaChainBuilder(t *testing.T, cfg *Config) *tastoradockertypes.Chai
 			Build()
 	}
 
+	// register the first wallet as the faucet wallet
+	addr, err := records[0].GetAddress()
+	require.NoError(t, err, "failed to get address from keyring record")
+	// Create the wallet with all required fields
+	faucetWallet := tastoradockertypes.NewWallet(addr, addr.String(), "celestia", records[0].Name)
```
Comment (chatton) [inline test/docker-e2e/dockerchain/testchain.go:59]: the implementation in tastora uses the faucet wallet to fund relayer wallets, we  can just use the first validator for this purpose.

File: test/docker-e2e/dockerchain/testchain.go [hunk 2]
```diff
@@ -51,15 +52,22 @@ func NewCelestiaChainBuilder(t *testing.T, cfg *Config) *tastoradockertypes.Chai
...
+
 	return tastoradockertypes.NewChainBuilder(t).
 		WithName("celestia"). // just influences home directory on the host.
 		WithChainID(cfg.Genesis.ChainID).
 		WithDockerClient(cfg.DockerClient).
 		WithDockerNetworkID(cfg.DockerNetworkID).
-		WithImage(tastoradockertypes.NewDockerImage(cfg.Image, cfg.Tag, "10001:10001")).
+		WithImage(tastoracontainertypes.NewImage(cfg.Image, cfg.Tag, "10001:10001")).
```
Comment (chatton) [inline test/docker-e2e/dockerchain/testchain.go:66]: some structural changes in tastora to split out the original mono-package

File: test/docker-e2e/e2e_ibc_test.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+// newSimappChainBuilder builds a standard simapp chain without token filters
+func newSimappChainBuilder(t *testing.T, cfg *dockerchain.Config) *tastoradockertypes.ChainBuilder {
+	encodingConfig := testutil.MakeTestEncodingConfig(app.ModuleEncodingRegisters...)
+	return tastoradockertypes.NewChainBuilder(t).
+		WithEncodingConfig(&encodingConfig).
+		WithName("simapp").
+		WithChainID("chain-b").
+		// TODO: this is a custom built simapp that has the bech32prefix as "celestia" as a workaround for the global
+		// SDK config not being usable when 2 chains have a different beck32 preix (e.g. "celestia" and "cosmos" ) because it is sealed.
+		WithImage(tastoracontainertypes.NewImage("ghcr.io/chatton/ibc-go-simd", "v8.5.0", "1000:1000")).
```
Comment (chatton) [inline test/docker-e2e/e2e_ibc_test.go:338]: this is the main thing that needs to be resolved before merging.

If we are happy to merge this for now we can create an issue to immediately move this image somewhere else or figure out a solution to the global sdk config.

File: test/docker-e2e/e2e_ibc_test.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+// This reuses the existing upgrade logic from e2e_upgrade_test.go
+func (s *CelestiaTestSuite) upgradeChain(ctx context.Context, chain tastoratypes.Chain, targetAppVersion uint64) {
+	validatorNode := chain.GetNodes()[0]
+	cfg := s.celestiaCfg
+	kr := cfg.Genesis.Keyring()
+
+	records, err := kr.List()
+	s.Require().NoError(err, "failed to list keyring records")
+
+	upgradeHeight := s.signalAndGetUpgradeHeight(ctx, chain, validatorNode, cfg, records, targetAppVersion)
```
Comment (chatton) [inline test/docker-e2e/e2e_ibc_test.go:264]: re-uses upgrade function from the upgrade tests.

File: test/docker-e2e/e2e_ibc_test.go [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+	s.transferTokens(ctx, chainA, chainB, channel, "utia", 100000)
+}
+
+// transferTokens performs an IBC token transfer from Celestia to simapp
+func (s *CelestiaTestSuite) transferTokens(ctx context.Context, sourceChain, destChain tastoratypes.Chain, channel ibc.Channel, denom string, amount int64) {
+	sourceWallet := sourceChain.GetFaucetWallet()
+	destWallet := destChain.GetFaucetWallet()
+	ibcTransfer := s.createIBCTransferMsg(sourceWallet, destWallet, channel, denom, amount)
+
+	// Submit transaction and verify results - always use TxClient since we only transfer from Celestia
```
Comment (chatton) [inline test/docker-e2e/e2e_ibc_test.go:200]: note: when we want to send a msg transfer back, we can't use the tx client as there are some hard coded assumptions we are dealing the celestia chain and it does not work generically with any other chains. Part of the reason to implement only transfer from celestia->chainb for now

File: test/docker-e2e/e2e_ibc_test.go (hunk 3 above)
Comment (evan-forbes) [reply test/docker-e2e/e2e_ibc_test.go:338]: LOL

**Review by evan-forbes — APPROVED** [review]
not blocking question

I'm not sure on the best way to index these tests with respect to upgrades. Yes we definitely want to test IBC, and there is an edge case where IBC breaks after upgrading but not when starting a chain from scratch. We are testing both here afaiu, but if we want to test all functionality after upgrading then maybe it makes sense to do all of that in the upgrade test and not upgrade in each e2e test 

🤷 still unsure

File: test/docker-e2e/e2e_ibc_test.go (hunk 3 above)
Comment (chatton) [reply test/docker-e2e/e2e_ibc_test.go:338]: that's a good way to sum up the situation :D

File: test/docker-e2e/go.mod [hunk 6]
```diff
@@ -270,11 +270,11 @@ require (
 
 replace (
 	cosmossdk.io/x/upgrade => github.com/celestiaorg/cosmos-sdk/x/upgrade v0.2.0
+	// goleveldb: canonical version
```
Comment (rootulp) [inline test/docker-e2e/go.mod]: [nit] the comment got moved to an incorrect place. IMO seems safe to remove this comment b/c it doesn't help maintainers

```suggestion
```

**Review by rootulp — APPROVED** [review]
